### ISSUE TITLE
Informer framework

### ISSUE THREAD

**Issue by sttts** [kubernetes/client-go#4]
To write third-party controllers having ~~object stores and~~ the informer framework would be very handy.

**Comment by sttts** [comment]
Object stores are included: https://github.com/kubernetes/client-go/tree/b7b3ea539440e6d0efc6ef4dab644108016788c0/1.4/tools/cache

**Comment by timothysc** [comment]
Many folks would like to refactor or eliminate the current informer entirely.

**Comment by deads2k** [comment]
> Many folks would like to refactor or eliminate the current informer entirely.

Can you elaborate on this?  I know that some people have suggested doing things differently, but what I've seen has only differed in implementation, nothing significant in the API interfaces.  I also haven't seen much action on it.

**Comment by wojtek-t** [comment]
Yeah - I'm also completely not aware of people that want to remove informer (though we obviously can improve it)

**Comment by timothysc** [comment]
@hongchaodeng and @xiang90 and proposed a refactoring to the informer in the 1.4 cycle that was pushed out.  I'd have to dig up the issue number.

**Comment by wongma7** [comment]
Is this the proposal in question? https://github.com/kubernetes/kubernetes/issues/27519

+1 to this anyway, I would rather use the same thing upstream controllers use, whatever flaws it has, than attempt to roll my own

**Comment by deads2k** [comment]
> Is this the proposal in question? kubernetes/kubernetes#27519
> 
> +1 to this anyway, I would rather use the same thing upstream controllers use, whatever flaws it has, than attempt to roll my own

That proposal isn't what kube uses.  Kube uses the `SharedInformers` as they exist today.  The proposal didn't interoperate well with the existing DeltaFifo structure that has done a reasonable job so far.

**Comment by lavalamp** [comment]
This was an oversight, we intended to include this library, too.

**Comment by lavalamp** [comment]
It looks like @mikedanese is moving it anyway, so it should show up soon.

**Comment by jimmidyson** [comment]
Any news on this? Chomping at the bit to migrate over to this & really would like the informer framework in place to simplify migration.

**Comment by mikedanese** [comment]
https://github.com/kubernetes/kubernetes/pull/32718 Merged. This should hit the 1.5 client once it's synced.

**Comment by jimmidyson** [comment]
So this client gets automatically synced from kube repo then? How often does that happen?

**Comment by lavalamp** [comment]
~ daily. But I need to approve a PR before the 1.5 client will show up here
:)

**Comment by seh** [comment]
If I'm looking to confine my Kubernetes dependencies to this repository, what's a good replacement for [`workqueue.Interface`](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/workqueue/queue.go#L23) from "k8s.io/kubernetes/pkg/util/workqueue”? 

Type [`cache.FIFO`](https://github.com/kubernetes/client-go/blob/master/1.4/tools/cache/fifo.go#L87) from “k8s.io/client-go/1.4/tools/cache” is similar, but I don’t see a way to unblock a caller waiting on `FIFO.Pop`, nor a way to shut the `FIFO` down. Is there a better choice for replacing `workqueue.Interface`?

**Comment by fabxc** [comment]
@lavalamp The 1.5/ sub directory is in but it doesn't include Informers. They are neither in the staging path in kubernetes/kubernetes.

Anything we could do to speed this up? It's blocking/complicating some high priority stuff for us.

**Comment by caesarxuchao** [comment]
The informer will show up after #33334 gets merged.

We need to manually run [staging/src/k8s.io/client-go/copy.sh](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/copy.sh) to populate changes to the staging area, then they will be published to client-go.

**Comment by lavalamp** [comment]
@caesarxuchao can we include the workqueue, as well?

**Comment by seh** [comment]
I've been working today on porting my code that relies on `workqueue` to use `cache.FIFO`, with some awkwardness, so if `workqueue` could come across I'd stick with it.

**Comment by lavalamp** [comment]
Yeah, workqueue and cache.FIFO are somewhat different animals-- we should
include workqueue.

**Comment by seh** [comment]
I see that #33334 merged today. Is it the case that the next run of _copy.sh_ will put the informer stuff into this repository?

**Comment by caesarxuchao** [comment]
#33334 ran the copy.sh. The next run (every 12 hours) of the bot will push the change to client-go. `workqueue` is still missing though.

**Comment by seh** [comment]
Thar she blows! Commit 5974905c2ab2b6ff24107cbf7bc397d770a675da includes [`cache.NewInformer`](https://github.com/kubernetes/client-go/blob/5974905c2ab2b6ff24107cbf7bc397d770a675da/1.5/tools/cache/controller.go#L211)—though only in version 1.5, [as hinted by @mikedanese earlier](https://github.com/kubernetes/client-go/issues/4#issuecomment-248641180).

**Comment by caesarxuchao** [comment]
> Should we expect it to arrive in the version 1.4 subtree soon, or is that out of scope?

Depending on how bad people want it. It requires cherrypicking https://github.com/kubernetes/kubernetes/pull/32718 and https://github.com/kubernetes/kubernetes/pull/33334, i.e., it requires some labor.

**Comment by mqliang** [comment]
@caesarxuchao  files under 1.5/tools/cache/ are in package testing, instead of cache, is it intended?

**Comment by caesarxuchao** [comment]
No, it's a bug. I'll take a look.

**Comment by seh** [comment]
@caesarxuchao, I'm not sure when it's safe for me to start using the 1.5 subtree. Is it guaranteed to work against a Kubernetes version 1.3 cluster? That hesitation is what had me using the 1.4 subtree.

**Comment by lavalamp** [comment]
The 1.5 subtree is backwards compatible. However, the go interface will not
be frozen until 1.5 is released, i.e., it's mutable--tracking head--until
then.

So if you use the 1.5 client, be aware that we may make interface changes
that require you to change your code. We haven't done this yet, but we
probably will.

I'm totally open to cherrypicking client changes into the 1.4 client
branch-- now that 1.4 is released we should be able to do that without
getting the evil eye from the release czar. Maybe after Chao fixes the
package name. :)

The thing we absolutely cannot do is change the interface in the 1.4
client--no changes there that will make folks who import it change their
code. We only make changes like that over version boundaries.

**Comment by seh** [comment]
Thank you for the explanation. I would appreciate a back-port, should one arrive soon, but in the meantime I'll try using the 1.5 client and see how it goes.

**Comment by jimmidyson** [comment]
> no changes there that will make folks who import it change their
code. We only make changes like that over version boundaries.

As you're versioning this in line with kubernetes itself, this is obviously goes against semver. Not that semver is a panacea but at least it gives a feel for API stability.

**Comment by lavalamp** [comment]
Indeed, that's a great point. Maybe we should version the go client
independently. I've put this on my list of things to think about.

**Comment by fabxc** [comment]
As it's pulled together from the kubernetes repo, independent versioning would ultimately leek into that. Even if intentionally developing on the most recent version, having the the sanitized package layout is valuable.

Maybe it's simplest to stick with freezing together with kubernetes but calling the non-frozen version "nightly" or "head"?

**Comment by lavalamp** [comment]
We talked a fair amount about this at the API Machinery SIG just now. I
think we definitely need to make it more clear what we're doing here.

Note that we haven't actually broken compatibility between 1.4 and 1.5 yet
so technically we're still following semver :)

**Comment by caesarxuchao** [comment]
I'll keep it open until we move workqueue.

**Comment by lavalamp** [comment]
I'm not quite sure why github closed this?

**Comment by caesarxuchao** [comment]
informers and workqueue are in the master branch now. Closing.
